keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,423 @@
+details in [user stories](#user-stories).
+
+### Goals
+
+- Even spreading is calculated among pods instead of apps API (such as
+  Deployment, ReplicaSet).
+- Even spreading can be either a predicate (hard requirement) or a priority
+  (soft requirement).
+
+### Non-Goals
+
+- Even spreading is NOT calculated on an application basis. In other words, it's
+  not only applied within replicas of an application, but also applied to
+  replicas of other applications if appropriate.
+- "Max number of pods per topology" is NOT a goal.
+- Scale-down on an application is not guaranteed to achieve even pods spreading
+  in the initial implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an application developer, I want my application pods to be scheduled onto
+specific topology domains as even as possible. Current status is that pods may
+be stacked onto a specific topology domain. (see
+[#68981](https://github.com/kubernetes/kubernetes/issues/68981))
+
+#### Story 2
+
+As an application developer, I want my application pods not to co-exist with
+specific pods (via PodAntiAffinity). But in some cases, it'd be favorable to
+tolerate "violating" pods in a manageable way. For example, suppose an app
+(replicas=2) is using PodAntiAffinity and deployed onto a 2-nodes cluster, and
+next the app needs to perform a rolling upgrade, then a third replacement pod is
+created, but it failed to be placed due to lack of resource. In this case,
+
+- if CA is enabled, a new machine will be provisioned to hold the new pod
+  (although old replicas will be deleted afterwards) (see
+  [#40358](https://github.com/kubernetes/kubernetes/issues/40358))
+- if CA is not enabled, it's a deadlock since the replacement pod can't be
+  placed. The only workaround at this moment is to update app strategyType from
+  "RollingUpdate" to "Recreate".
+
+Neither of them is an ideal solution. A promising solution is to give user an
+option to trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
+
+### Risks and Mitigations
+
+The feature requires additional processing for pods that use it and it is ok to
+have some performance overhead. But we will make sure our implementation will
+not have any performance penalty for pods that do not use this feature.
+
+## Design Details
+
+### API
+
+A new structure called `TopologySpreadConstraint` is introduced which acts as a
+standalone spec and is applied to `pod.spec`. It's only effective when it's not
+nil.
+
+```go
+type PodSpec struct {
+    // TopologySpreadConstraint describes how a group of pods are spread
+    // If specified, scheduler will enforce the constraint
+    // +optional
+    TopologySpreadConstraint *TopologySpreadConstraint
+    ......
```

Comment:
lets show other interesting fields of PodSpec here like ` Affinity *Affinity` and `NodeSelector` and then other ideas will pop up in peoples mind